File: 1-working-with-list/p21_insert_at.ml

let rec insert_at e p = function
  | [] ->
    if p = 0 then
      [e]
    else
      []
  | (h :: t) as l ->
    if p = 0 then
      e :: l
    else
      h :: (insert_at e (p-1) t);;

insert_at "alfa" 1 ["a";"b";"c";"d"];;
insert_at "alfa" 3 ["a";"b";"c";"d"];;
insert_at "alfa" 4 ["a";"b";"c";"d"];;

let insert_at e p lst =
  let rec aux acc curr = function
    | [] ->
      if curr = 0 then
        List.rev (e::acc)
      else
        List.rev acc
    | (h :: t) as l ->
      if curr = 0 then
        List.rev (e::acc) @ l
      else
        aux (h::acc) (curr-1) t
  in aux [] p lst;;

insert_at "alfa" 1 ["a";"b";"c";"d"];;
insert_at "alfa" 3 ["a";"b";"c";"d"];;
insert_at "alfa" 4 ["a";"b";"c";"d"];;

File: 1-working-with-list/p21_insert_at.scala
object Main {

  def main(args: Array[String]) {
    val l = List("a", "b", "c", "d")
    println(insertAt("alpha", 1, l))
    println(insertAt("alpha", 0, l))
    println(insertAt("alpha", 3, l))
    println(insertAt("alpha", 4, l))
    println(insertAt("alpha", 6, l))
  }

  def insertAt[T](e: T, index: Int, list: List[T]): List[T] = {
    def aux(l: List[T], c: Int, acc: List[T]): List[T] = {
      l match {
        case Nil => (e :: acc).reverse
        case hd :: tail =>
          if(c == 0)
            acc.reverse ::: (e :: l)
          else
            aux(tail, c-1, hd :: acc)
      }
    }
    aux(list, index, List())
  }
}
